tensorboard/loader.py

```diff
@@ -0,0 +1,414 @@
+      Exception: To propagate any exceptions that may have been thrown
+          by the read operation in the other thread. If an exception is
+          thrown, then all subsequent calls to this method will rethrow
+          that same exception.
+
+    :rtype: Record
+    """
+    with self._lock:
+      if self._is_closed:
+        raise IOError('%s is closed' % self)
+      if not self._is_started:
+        self._thread.start()
+        self._is_started = True
+      else:
+        record = self._get_record()
+        if record is not None:
+          return record
+        self._is_tapped = False
+        self._please_sir_i_want_some_more.notify()
+      while (self._read_exception is None and
+             not self._is_tapped and
+             not self._records):
+        self._no_soup_for_you.wait()
+      return self._get_record()
+
+  def close(self):
+    """Closes event log reader if open.
+
+    If the i/o thread is running, this method blocks until it has been
+    shut down.
+
+    Further reads are not permitted after this method is called.
+
+    Raises:
+      Exception: To propagate any exceptions that may have been thrown
+          by the close operation in the other thread. If an exception
+          is thrown, then all subsequent calls to this method will
+          rethrow that same exception.
+    """
+    with self._lock:
+      if not self._is_closed:
+        self._is_closed = True
+        if not self._is_started:
+          return
+        self._please_sir_i_want_some_more.notify()
+      while self._reader is not None:
+        self._no_soup_for_you.wait()
+      if self._close_exception is not None:
+        raise self._close_exception
+
+  def _get_record(self):
+    if self._read_exception is not None:
+      raise self._read_exception
+    if not self._records:
+      return None
+    record = self._records.popleft()
+    self._buffered -= len(record.record)
+    if self._should_wakeup():
+      self._please_sir_i_want_some_more.notify()
+    return record
+
+  def _run(self):
+    while True:
+      with self._lock:
+        while not self._should_wakeup():
+          self._please_sir_i_want_some_more.wait()
+        if self._is_closed:
+          try:
+            self._reader.close()
+            tf.logging.debug('Closed')
+          except Exception as e:
+            self._close_exception = e
+            tf.logging.debug('Close failed: %s', e)
+          self._reader = None
+          self._no_soup_for_you.notify_all()
+          return
+        if not self._should_rebuffer():
+          continue
+        # Calculate a good amount of data to read outside the lock.
+        # The less we have buffered, the less re-buffering we'll do.
+        # We want to minimize wait time in the other thread.
+        x = float(self._buffered)
+        c = float(self._read_ahead)
+        s = BufferedRecordReader.READ_AHEAD_AGGRESSION
+        want = int(min(c - x, s/c * x**s + 1))
```

Comment:
Added more comment information.

I added some more tests too. But I'm not quite sure how to test much further than this in Python. Normally to test something like this you need [Phaser](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Phaser.html).